### PR TITLE
Bug/5236 - Add Retry button follow-up

### PR DESCRIPTION
**Pull request by hussain-t**
## Summary

Addresses issue:

- #5236 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/components/ReportError.test.js [hunk 1]
```diff
@@ -234,4 +282,38 @@ describe( 'ReportError', () => {
 
 		expect( queryByText( /retry/i ) ).toBeInTheDocument();
 	} );
+
+	it( "should dispatch the `invalidateResolution` action the equal times of errors' length", () => {
```
Comment (tofumatt) [inline assets/js/components/ReportError.test.js:286]:
```suggestion
	it( "should dispatch the `invalidateResolution` action for each retry-able error", () => {
```

This test should also be expanded to ensure only errors that can be retried, eg. https://github.com/google/site-kit-wp/pull/5492/files#diff-b95a49b3fe9c5fd7c5b5f782eef76c5c71fd74299f4bdce4a7156144db48d621R111-R118 are invalidated and called. Adding an extra error that cannot be retried would be good in this test.

File: assets/js/components/ReportError.js [hunk 2]
```diff
@@ -51,43 +52,81 @@ export default function ReportError( { moduleSlug, error } ) {
...
+		title = sprintf(
+			/* translators: %s: module name */
+			__( 'Data errors in %s', 'google-site-kit' ),
+			module?.name
+		);
+	} else {
 		title = sprintf(
 			/* translators: %s: module name */
-			__( 'Insufficient permissions in %s', 'google-site-kit' ),
+			__( 'Data error in %s', 'google-site-kit' ),
```
Comment (tofumatt) [inline assets/js/components/ReportError.js:68]: Sorry, why this change?

File: assets/js/components/ReportError.js [hunk 3]
```diff
@@ -51,43 +52,81 @@ export default function ReportError( { moduleSlug, error } ) {
...
 	const dispatch = useDispatch();
 
-	let title = sprintf(
-		/* translators: %s: module name */
-		__( 'Data error in %s', 'google-site-kit' ),
-		module?.name
-	);
-	let message = error.message;
+	const errors = Array.isArray( error ) ? error : [ error ];
+	const uniqueErrors = uniqueBy( errors, ( e ) => e.code && e.message );
```
Comment (tofumatt) [inline assets/js/components/ReportError.js:56]: It's worth adding a test for this behaviour as well.

**Review by tofumatt — CHANGES_REQUESTED** [review]
I think it's worth adding a few more tests to this behaviour before merging (and I have a few follow-up questions), but the approach looks right to me.

I'll be out for the rest of the day so best to assign follow-up PRs to either @techanvil or @felixarntz 🙂

File: assets/js/components/ReportError.js (hunk 2 above)
Comment (hussain-t) [reply assets/js/components/ReportError.js:68]: @tofumatt, if there are more than one unique/different errors, it's better to display `errors` instead of `error` in the title.

If you ask why `Insufficient permissions in %s` is changed to `Data error in %s`, that is due to the line changes GitHub shows. However, the logic remains the same.

File: assets/js/components/ReportError.js (hunk 3 above)
Comment (techanvil) [inline assets/js/components/ReportError.js:56]: This doesn't look quite right, as it's not testing the combined equality of both `code` and `error` (due to the way `&&` short circuits only a falsy value of `code` _or_ the value of `message` will be used as the unique value), and it's also not taking `reconnectURL` into account.

I think this would probably be the right approach:

```js
import uniqWith from 'lodash/uniqWith`;

// ...

const uniqueErrors = uniqWith( errors, ( errorA, errorB ) =>
  errorA.code === errorB.code &&
  errorA.message === errorB.message &&
  errorA.data?.reconnectURL === errorB.data?.reconnectURL
);
```

File: assets/js/components/ReportError.js [hunk 4]
```diff
@@ -51,43 +52,81 @@ export default function ReportError( { moduleSlug, error } ) {
...
+	};
+
+	const description = (
+		<Fragment>
+			{ uniqueErrors.map( ( e ) => {
+				const message = getMessage( e );
+				const reconnectURL = error?.data?.reconnectURL;
+				return reconnectURL ? (
+					<ErrorText
+						key={ e.message }
```
Comment (techanvil) [inline assets/js/components/ReportError.js:98]: I don't think `message` is a good choice of key here, as there's a chance that multiple errors may occur with the same `message` but a different `code` or `reconnectURL`. Bit of an edge case, but the key should either reflect the set of values that makes an error unique (i.e. `code`, `message` and `reconnectURL`) - or simply use `index`...

**Review by techanvil — CHANGES_REQUESTED** [review]
Nice one @hussain-t. I have added a couple of comments around the uniqueness of errors.

File: assets/js/components/ReportError.js (hunk 3 above)
Comment (techanvil) [reply assets/js/components/ReportError.js:56]: Or (as discussed "offline") maybe something like this would be more accurate -

```js
const uniqueErrors = errors
  .map( e => ( { message: getMessage( e.message ), reconnectURL: e.data?.reconnectURL }) )
  .uniqWith( (errorA, errorB) =>
    errorA.message === errorB.message &&
    errorA.reconnectURL === errorB.reconnectURL
  );
```

Comment (hussain-t) [reply assets/js/components/ReportError.js:56]: Thanks, @techanvil, yes, this approach is good except that passing the mapped errors array as the first argument to `uniqWith`:

```js
const uniqueErrors = uniqWith(
	errors.map( ( e ) => ( {
		message: getMessage( e ),
		reconnectURL: e.data?.reconnectURL,
	} ) ),
	( errorA, errorB ) =>
		errorA.message === errorB.message &&
		errorA.reconnectURL === errorB.reconnectURL
);
```

Comment (techanvil) [reply assets/js/components/ReportError.js:56]: Thanks @hussain-t, good spot!

**Review by tofumatt — APPROVED** [review]
Looks good, and when I tested locally with the AdSense error raised by @felixartnz here: https://github.com/google/site-kit-wp/issues/5236#issuecomment-1171509120 it now worked.

Great work! 👍🏻
